pkg/controllers/inventory/inventory_controller.go

```diff
@@ -0,0 +1,216 @@
+}
+
+func watchPod(c *InventoryController, mgr ctrl.Manager) error {
+	podInformer, err := mgr.GetCache().GetInformer(context.Background(), &v1.Pod{})
+	if err != nil {
+		log.Error(err, "Create pod Informer error")
+		return err
+	}
+
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			// Handle Pod add event
+			c.handlePod(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			// Handle Pod update event
+		},
+		DeleteFunc: func(obj interface{}) {
+			// Handle Pod delete event
+			c.handlePod(obj)
+		},
+	})
+	return nil
+}
+
+func (c *InventoryController) Run(stopCh <-chan struct{}) {
+	defer c.inventoryObjectQueue.ShutDown()
+	log.Info("Starting inventory controller")
+
+	// The sync state of Gateway informer should not be checked here, since no other inventory resource types rely on
+	// Gateway store, and the Gateway informer readiness is polled and started in background.
+	log.Info("Waiting for caches to sync for inventory controller")
+
+	log.Info("Caches are synced for Inventory controller, cleaning up stale objects")
+	err := c.CleanStaleInventoryObjects()
+	if err != nil {
+		log.Error(err, "Failed to clean up stale inventory objects")
+		return
+	}
+
+	// Batch update inventory based on batch time and size.
+	// Inventory worker will be running in forever loop util inventoryMutex is locked by inventoryTimeWorker.
+	// Only one worker processes and sends request to NSX MP at one time.
+	go wait.Until(c.inventoryWorker, time.Second, stopCh)
+	go wait.Until(c.inventoryTimeWorker, time.Second*time.Duration(c.cf.InventoryBatchPeriod), stopCh)
+	go wait.JitterUntil(c.inventoryGCWorker, commonservice.GCInterval, inventoryGCJitterFactor, true, stopCh)
+	<-stopCh
+}
+
+func (c *InventoryController) inventoryTimeWorker() {
+	defer c.inventoryMutex.Unlock()
+
+	c.inventoryMutex.Lock()
+	if len(c.keyBuffer) > 0 {
+		c.syncInventoryKeys()
+	}
+}
+
+func (c *InventoryController) inventoryGCWorker() {
+	defer c.inventoryMutex.Unlock()
+
+	c.inventoryMutex.Lock()
+	if err := c.CleanStaleInventoryObjects(); err != nil {
+		log.Error(err, "Failed to garbage collect inventory resources")
+	}
+}
+
+func (c *InventoryController) inventoryWorker() {
+	for c.processNextInventoryWorkItem() {
+	}
+}
+
+func (c *InventoryController) processNextInventoryWorkItem() bool {
+	key, quit := c.inventoryObjectQueue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.inventoryMutex.Unlock()
+	c.inventoryMutex.Lock()
+	c.keyBuffer.Insert(key.(inventory.InventoryKey))
+	if len(c.keyBuffer) >= c.cf.InventoryBatchSize {
+		c.syncInventoryKeys()
+	}
+	return true
+}
+
+func (c *InventoryController) syncInventoryKeys() {
+	// Remove all the keys from processing and clear keyBuffer.
+	defer func() {
+		for key := range c.keyBuffer {
+			c.inventoryObjectQueue.Done(key)
+		}
+		c.keyBuffer = inventory.KeySet{}
+	}()
+
+	if len(c.keyBuffer) >= 0 {
+		retryKeys, err := c.service.SyncInventoryObject(c.keyBuffer)
```

Comment:
why not return failedKeys and succeedKeys directly, then you don't need to check retryKeys.Has(key)

Reply:
check the key to avoid the same (key)request in the queue